onnxruntime/contrib_ops/rocm/fused_conv.cc

```diff
@@ -0,0 +1,321 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <unordered_set>
+#include "core/providers/rocm/nn/conv.h"
+#include "core/providers/rocm/rocm_common.h"
+
+namespace onnxruntime {
+namespace contrib {
+namespace rocm {
+
+namespace {
+
+// Copied from hipDNN/library/src/hcc_detail/hipdnn_miopen.cpp
+miopenStatus_t _miopenAddTensor(
+    miopenHandle_t handle,
+    const void *alpha,
+    const miopenTensorDescriptor_t aDesc,
+    const void *A,
+    const void *beta,
+    const miopenTensorDescriptor_t cDesc,
+    void *C,
+    const void* zero_scalar)
+{
+    const miopenTensorOp_t tensorOp = miopenTensorOpAdd;
+    // opnd2 = Add ( 0.0 * opnd0, alpha * opnd1 ) + alpha * opnd2
+    return miopenOpTensor(handle, tensorOp,
+                          zero_scalar, cDesc, C,
+                          alpha, aDesc, A,
+                          alpha, cDesc, C);
+}
+
+}
+
+template <typename T>
+class FusedConv : public onnxruntime::rocm::Conv<T> {
+ public:
+  using Base = onnxruntime::rocm::Conv<T>;
+  FusedConv(const OpKernelInfo& info) : onnxruntime::rocm::Conv<T>(info) {
+    std::string activation;
+    if (info.GetAttr<std::string>("activation", &activation) == Status::OK() &&
+        MapMode(activation) == Status::OK() &&
+        miopenCreateActivationDescriptor(&activation_desc_) == miopenStatusSuccess) {
+      status_ = miopenSetActivationDescriptor(activation_desc_,
+                                              activation_mode_,
+                                              0.0, 0.0, 0.0);
+    }
+  }
+
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(FusedConv);
+
+  ~FusedConv() {
+    if (activation_desc_) {
+      miopenDestroyActivationDescriptor(activation_desc_);
+      status_ = miopenStatusNotInitialized;
+      activation_desc_ = nullptr;
+    }
+  }
+
+  Status ComputeInternal(OpKernelContext* context) const override {
+    MIOPEN_RETURN_IF_ERROR(status_);
+    std::lock_guard<OrtMutex> lock(Base::s_.mutex);
+
+    // Detect if the dimesion changed
+    const Tensor* X = context->Input<Tensor>(0);
+    const TensorShape& x_shape = X->Shape();
+    const auto x_dims = x_shape.GetDims();
+    bool input_dims_changed = (Base::s_.last_x_dims.GetDims() != x_dims);
+
+    const Tensor* W = context->Input<Tensor>(1);
+    const TensorShape& w_shape = W->Shape();
+    auto w_dims = w_shape.AsShapeVector();
+    bool w_dims_changed = (Base::s_.last_w_dims.GetDims() != gsl::make_span(w_dims));
+
+    ORT_RETURN_IF_ERROR(Base::UpdateState(context, true));
+    if (Base::s_.Y->Shape().Size() == 0) {
+      return Status::OK();
+    }
+
+    bool has_z = nullptr != Base::s_.z_data;
+    bool has_b = nullptr != Base::s_.b_data;
+
+    // TODO: Can be removed once MIOpen supports multiple Bias operators in
+    //       single Fusion plan.
+    bool should_try_fusion_api = true;
+    if (has_z && has_b && this->unsupported_fusion_both_z_b_) {
```

Comment:
This actually reminds me the cache thing for Conv implementation. For a general solution, is it possible to maintain a global map/cache from MiopenConvState to FusedConvFusionData as well as success flag (init to unknown)? So for each different MiopenConvState, we look into the cache, if flag is unknown, try to create one, if not, then just return the existing one or nullptr if any error before. We don't need to limit the failure case for 2-bias case only. Maybe different ROCm version has different failure cases and this general cache should work for all such cases.

Reply:
I was thinking about this as well, but realized it is non-trivial to cache all input arguments.
The pseudo code may look like
```
def ComputeFusedConv(self, args):
    key = hash(args) # <- Non Trivial part, and error prone due to missing arguments in the hashing.
    if key in self._failure_cache:
        skip_fusion = True
    else:
        skip_fusion = False
```

This mechanism, however, can be reused to implement a more efficient Fusion API backend where the fusion plan may be shared among FusedConv Ops with identical configurations. (I left a comment about this scheme in the previous commit)

Maybe a better approach is to have a global factory for Fusion Plan objects so that more sophisticated logic can be put there and all future Fusion API users can benefit from it.

Reply:
Certainly we can.

The challenge here is the parameters required to setup a Fusion Plan is not POD.
The biggest problem is miopenTensorDescriptor_t (implicitly converted from, e.g., `Base::s_.x_tensor`), whose underlying datatype is actually `void*`.
The hashing process will be verbose and error-prone because calling Get functions is essential to retrieve the real parameters.

Let me give it a try and see if worth it.

Reply:
any update on this?

Reply:
Finally have some time  to work on this. Updates coming in a few hours.